core/client/src/main/java/alluxio/client/block/stream/BlockInStream.java

```diff
@@ -0,0 +1,188 @@
+   * @throws IOException if it fails to create an instance
+   * @return the {@link BlockInStream} created
+   */
+  public static BlockInStream createLocalBlockInStream(long blockId, long blockSize,
+      WorkerNetAddress workerNetAddress, BlockStoreContext context, InStreamOptions options)
+      throws IOException {
+    Closer closer = Closer.create();
+
+    try {
+      BlockWorkerClient blockWorkerClient =
+          closer.register(context.createWorkerClient(workerNetAddress));
+      LockBlockResult lockBlockResult = blockWorkerClient.lockBlock(blockId);
+      LocalFileBlockReader localFileBlockReader =
+          new LocalFileBlockReader(lockBlockResult.getBlockPath());
+      PacketReader.Factory factory = new LocalFilePacketReader.Factory(localFileBlockReader);
+      return new BlockInStream(factory, blockId, blockSize, blockWorkerClient, options);
+    } catch (AlluxioException e) {
+      closer.close();
+      throw new IOException(e);
+    } catch (IOException e) {
+      closer.close();
+      throw e;
+    }
+  }
+
+  /**
+   * Creates an instance of remote {@link BlockInStream} that reads from a remote worker.
+   *
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param workerNetAddress the worker network address
+   * @param context the block store context
+   * @param options the options
+   * @throws IOException if it fails to create an instance
+   * @return the {@link BlockInStream} created
+   */
+  public static BlockInStream createRemoteBlockInStream(long blockId, long blockSize,
+      WorkerNetAddress workerNetAddress, BlockStoreContext context, InStreamOptions options)
+    throws IOException {
+    Closer closer = Closer.create();
+
+    try {
+      BlockWorkerClient blockWorkerClient =
+          closer.register(context.createWorkerClient(workerNetAddress));
+      LockBlockResult lockBlockResult = blockWorkerClient.lockBlock(blockId);
+      PacketReader.Factory factory = new NettyPacketReader.Factory(
+          blockWorkerClient.getDataServerAddress(), blockId, lockBlockResult.getLockId(),
+          blockWorkerClient.getSessionId());
+      return new BlockInStream(factory, blockId, blockSize, blockWorkerClient, options);
+    } catch (AlluxioException e) {
+      closer.close();
+      throw new IOException(e);
+    } catch (IOException e) {
+      closer.close();
+      throw e;
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    try {
+      mBlockWorkerClient.unlockBlock(mBlockId);
+    } catch (Throwable e) { // must catch Throwable
+      throw mCloser.rethrow(e); // IOException will be thrown as-is
+    } finally {
+      mCloser.close();
+    }
+  }
+
+  @Override
+  public long remaining() {
+    return mInputStream.remaining();
+  }
+
+  @Override
+  public void seek(long pos) throws IOException {
+    mInputStream.seek(pos);
+  }
+
+  @Override
+  public int read(long pos, byte[] b, int off, int len) throws IOException {
```

Comment:
the general rule of overloading method is they should be exactly the same, except the parameter(s) can be different.
In this case, I view `InputStream.read` works differently from `PositionedReadable.read`, as one moves position while the other not. This difference is significant. And I think keeping our signature consistent with Hadoop signature is not really important unless I missed anything. So I incline to keep this signature different from `read`. e.g., `positionedRead` ?